yt_dlp/extractor/youtube.py

```diff
@@ -584,8 +587,200 @@ def _real_initialize(self):
+            'expiry': time_seconds(
+                seconds=traverse_obj(token_response, ('expires_in', {float_or_none}), default=300) - 10),
+        })
+        refresh_token = traverse_obj(token_response, 'refresh_token', {str})
+        if refresh_token:
+            self.cache.store(self._NETRC_MACHINE, self._get_oauth_cache_key(profile), refresh_token)
+            YoutubeBaseInfoExtractor._OAUTH_ACCESS_TOKEN_CACHE[profile]['refresh_token'] = refresh_token
+
+    def _initialize_oauth(self, user, refresh_token):
+        self._OAUTH_PROFILE = user or 'default'
+
+        if self._OAUTH_PROFILE in YoutubeBaseInfoExtractor._OAUTH_ACCESS_TOKEN_CACHE:
+            self.write_debug(f'{self._OAUTH_DISPLAY_ID}: Using cached access token for profile "{self._OAUTH_PROFILE}"')
+            return
+
+        YoutubeBaseInfoExtractor._OAUTH_ACCESS_TOKEN_CACHE[self._OAUTH_PROFILE] = {}
+
+        if refresh_token == "''":
+            refresh_token = None
+
+        # Allow refresh token passed to initialize cache
+        if refresh_token:
+            self.cache.store(self._NETRC_MACHINE, self._get_oauth_cache_key(self._OAUTH_PROFILE), refresh_token)
+
+        refresh_token = refresh_token or self.cache.load(self._NETRC_MACHINE, self._get_oauth_cache_key(self._OAUTH_PROFILE))
+        if refresh_token:
+            YoutubeBaseInfoExtractor._OAUTH_ACCESS_TOKEN_CACHE[self._OAUTH_PROFILE]['refresh_token'] = refresh_token
+            try:
+                token_response = self._refresh_token(refresh_token)
+            except ExtractorError as e:
+                self.report_warning(f'{self._OAUTH_DISPLAY_ID}: Failed to refresh access token: {e}')
+                token_response = self._oauth_authorize
+        else:
+            token_response = self._oauth_authorize
+
+        self._set_oauth_info(token_response, self._OAUTH_PROFILE)
+        self.write_debug(f'{self._OAUTH_DISPLAY_ID}: Logged in using profile "{self._OAUTH_PROFILE}"')
+
+    def _refresh_token(self, refresh_token):
+        try:
+            token_response = self._download_json(
+                self._OAUTH_TOKEN_ENDPOINT,
+                video_id=self._OAUTH_DISPLAY_ID,
+                note='Refreshing access token',
+                data=json.dumps({
+                    'client_id': self._OAUTH_CLIENT_ID,
+                    'client_secret': self._OAUTH_CLIENT_SECRET,
+                    'refresh_token': refresh_token,
+                    'grant_type': 'refresh_token',
+                }).encode(),
+                headers={'Content-Type': 'application/json'})
+        except ExtractorError as e:
+            if isinstance(e.cause, HTTPError):
+                token_response = self._parse_json(
+                    self._webpage_read_content(e.cause.response, None, self._OAUTH_DISPLAY_ID) or '{}',
+                    video_id=self._OAUTH_DISPLAY_ID)
+                error = traverse_obj(token_response, 'error', {str})
```

Comment:
passing the url might be nice for `--write-pages` purposes, and we can use traversal and `fatal=False` to make it totally non-fatal
```suggestion
                error = traverse_obj(
                    self._webpage_read_content(e.cause.response, self._OAUTH_TOKEN_ENDPOINT, self._OAUTH_DISPLAY_ID, fatal=False),
                    ({json.loads}, 'error', {str}))
```